clientlibrary/worker/worker.go

```diff
@@ -324,6 +375,90 @@ func (w *Worker) eventLoop() {
 	}
 }
 
+func (w *Worker) rebalance() error {
+	log := w.kclConfig.Logger
+
+	workers, err := w.leasestealer.ListActiveWorkers(w.shardStatus)
+	if err != nil {
+		log.Debugf("Error listing workers. workerID: %s. Error: %+v ", w.workerID, err)
+		return err
+	}
+
+	// Only attempt to steal one shard at at time, to allow for linear convergence
+	if w.shardStealInProgress {
+		shardInfo := make(map[string]bool)
+		err := w.getShardIDs("", shardInfo)
+		if err != nil {
+			return err
+		}
+		for _, shard := range w.shardStatus {
+			if shard.ClaimRequest != "" && shard.ClaimRequest == w.workerID {
+				log.Debugf("Steal in progress. workerID: %s", w.workerID)
+				return nil
+			}
+			// Our shard steal was stomped on by a Checkpoint.
+			// We could deal with that, but instead just try again
+			w.shardStealInProgress = false
+		}
+	}
+
+	var numShards int
+	for _, shards := range workers {
+		numShards += len(shards)
+	}
+
+	numWorkers := len(workers)
+
+	// 1:1 shards to workers is optimal, so we cannot possibly rebalance
+	if numWorkers >= numShards {
```

Comment:
Assume there are 6 shards and 3 workers. Each holds 2 shards. Adding more workers will always keep those additional workers idle since optimal shard count has been reached.

Reply:
I don't think that's true. See: https://github.com/connormckelvey/vmware-go-kcl/blob/lease-stealing-support/clientlibrary/worker/worker.go#L422

A worker without shards WILL eventually steal one. At which point it will be represented in dynamodb and the next time optimal shards per worker is recalculated by each worker, it will equal 1 (via integer division).

Reply:
ActiveWorkers are those shown inside dynamodb table and all of those hold at least 1 shard. The idle worker is not inside dynamodb table. I think it will be much easier to create a real stream on AWS and test those scenarios.